### PR TITLE
[InstCombine] Fix behavior for `(fmul (sitfp x), 0)`

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[InstCombine] Add test for `(fmul (sitfp x), 0)`; NFC**
- **[InstCombine] Fix behavior for `(fmul (sitfp x), 0)`**

Bug was introduced in #82555

We where missing check that the constant was non-zero for signed + mul
transform.

### CODE REVIEW

**Review by nikic — CHANGES_REQUESTED** [review]
I haven't checked whether it might be safe in this context, but the m_NonZero you added here is an unacceptable footgun: It allows undef values that may refine to zero.

**Comment by goldsteinn** [comment]
> I haven't checked whether it might be safe in this context, but the m_NonZero you added here is an unacceptable footgun: It allows undef values that may refine to zero.

This matches `m_NonZeroFp` (Which I guess I could have just used instead).

File: llvm/test/Transforms/InstCombine/binop-itofp.ll [hunk 1]
```diff
@@ -1004,3 +1004,31 @@ define float @test_ui_add_with_signed_constant(i32 %shr.i) {
...
+  %sel = select i1 %.b, i32 65529, i32 53264
+  %conv.i = trunc i32 %sel to i16
+  %conv1.i = sitofp i16 %conv.i to float
+  %mul3.i.i = fmul float %conv1.i, 0.000000e+00
+  store i32 %sel, ptr %g_2345, align 4
+  %a.0.copyload.cast = bitcast float %mul3.i.i to i32
+  %cmp = icmp sgt i32 %a.0.copyload.cast, -1
+  %conv = zext i1 %cmp to i32
+  ret i32 %conv
+}
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/binop-itofp.ll:1034]: This test does not look minimal.

Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/binop-itofp.ll:1034]: This is from `llvm-reduce`.

The volatile store/ptrs where necessary to hit the bug (tested). Think the issue is `fmul x, 0.0` tends to be simplified on its own.

Comment (nikic) [reply llvm/test/Transforms/InstCombine/binop-itofp.ll:1034]: Can't you reduce it to at least this?
```
define float @missed_nonzero_check_on_constant_for_si_fmul(i1 %c, i1 %.b, ptr %g_2345) {
  %sel = select i1 %c, i32 65529, i32 53264
  %conv.i = trunc i32 %sel to i16
  %conv1.i = sitofp i16 %conv.i to float
  %mul3.i.i = fmul float %conv1.i, 0.000000e+00
  store i32 %sel, ptr %g_2345, align 4
  ret float %mul3.i.i
}
```

Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/binop-itofp.ll:1034]: let me test.

**Comment by goldsteinn** [comment]
> > I haven't checked whether it might be safe in this context, but the m_NonZero you added here is an unacceptable footgun: It allows undef values that may refine to zero.
> 
> This matches `m_NonZeroFp` (Which I guess I could have just used instead).

Made default no-undef + tests.

**Comment by goldsteinn** [comment]
> > > I haven't checked whether it might be safe in this context, but the m_NonZero you added here is an unacceptable footgun: It allows undef values that may refine to zero.
> > 
> > 
> > This matches `m_NonZeroFp` (Which I guess I could have just used instead).
> 
> Made default no-undef + tests.

If you would prefer, I can just drop the whole `m_NonZero` stuff and use `m_NonZeroFP`.

**Comment by nikic** [comment]
Yes, please drop it.

**Comment by goldsteinn** [comment]
> Yes, please drop it.

Done

**Review by nikic — APPROVED** [review]
LGTM

**Pull request closed without merging**